### ISSUE TITLE
ALNS parameter optimization

### ISSUE THREAD

**Issue by dietmarwo** [N-Wouda/ALNS#170]
#### Is your feature request related to a problem? Please describe
Checking out the examples, for some of them (TSP, CVRP) I found that using decay = 0.85 improved the results significantly compared to the choosen value decay = 0.8. A useful feature would be an automated support for parameter testing. 

#### Describe the solution you'd like
The solution would be an automated parameter swipe executing many optimization runs thereby changing one or more parameters
and the random seed. To utilize modern many-core CPUs these runs could be parallelized using Python multi-processing. 
Result would be mean and standard deviation of the objective for specific parameter values which could be visualized using plots. 

#### Describe alternatives you have considered
Alternative would be some guidance how to manually optimize the parameters - may be another example or based on an existing one.
Permutation Flow Shop already contains a section "Tuning the destroy rate". There could be a similar section for TSP called "Tuning the decay rate".

**Comment by N-Wouda** [comment]
Hi @dietmarwo! 👋 How does this compare to the ideas laid out in #109?

**Comment by dietmarwo** [comment]
Not sure whether ML hyperperameter tuning is optimial here. In ML we usually use dedicated hardware (TPUs/GPUs) which blocks parallel evaluation of hyperparameter settings. ALNS is often applied to objective functions which can be evaluated single threaded, so on a modern CPU you could run >= 16 experiments in parallel. But nevertheless, you are right, applying these existing tools is better than nothing.

**Comment by N-Wouda** [comment]
I have had some success applying [SMAC3](https://github.com/automl/SMAC3) to tune an ALNS before - some example code is available [here](https://github.com/N-Wouda/PL-Hourly-Heuristic), in `src/tune.py`. 

Alternatively:
- We could add more to the docs about how to run these experiments using a hand-rolled tuner (like we did for the flow shop example). This'd have to be fast, showing just the general ideas but not performing a complete, many-hours-long optimisation run.
- Find some other tuner? I have used SMAC for algorithm configuration, and also briefly looked at irace and ParamILS in the past. There might be alternatives that suit ALNS better: [this survey](https://arxiv.org/abs/2202.01651) seems like a promising start to finding those.

**Comment by dietmarwo** [comment]
Indeed, in SMAC3 "Scenario" supports the "n_jobs" parameter where you also can execute multiple runs in parallel. It may indeed be a good solution. Will also have a look at [this survey](https://arxiv.org/abs/2202.01651) .

**Comment by dietmarwo** [comment]
Related to the parallelization topic: 
I am currently thinking how to add parallelization to ALNS itself. For instance I could implement a "parallel_repair" operator which applies several "repairs" in parallel and submits only the one giving the best objective value. But this would probably be inferior to real parallelism implemented into the algorithm itself: Parallel destroys/repairs would use all repair results, not just the best one. Do you have any thoughts / plans about this?

**Comment by N-Wouda** [comment]
We used to have an issue open about just that, here: #59. I think this is the most basic parallel ALNS paper out there: [PALNS](https://backend.orbit.dtu.dk/ws/portalfiles/portal/4129408/PALNS.pdf). Whether that's the best way to do it is an open question. There are some papers from the past few years that describe parallel ALNS implementations, but I've not seen any that really convinced me they are significantly better than the obvious parallellisation of doing $N$ independent ALNS runs in parallel, and returning the best solution out of those. I don't think I can help you much in this direction.

**Comment by N-Wouda** [comment]
I think something like that should be possible, but I don't have the time right now to mock something up. It might be tricky to do this without breaking existing interfaces. Would you be willing to have a go at this if you're interested?
